nexus/reconfigurator/planning/src/blueprint_builder/external_networking.rs

```diff
@@ -149,6 +178,38 @@ impl<'a> BuilderExternalNetworking<'a> {
             }
         }
 
+        // Recycle the IP addresses of expunged or about-to-be-expunged
+        // external DNS zones. The latter is an unfortunate consequence
+        // of when this builder is constructed. An alternative would be
+        // a hook in expungement, but that seems possibly worse than this
+        // (admittedly awkward) filter.
+        // TODO: Remove when external DNS addresses come from policy.
```

Comment:
If I'm understanding correctly, the problem this filter is trying to deal with is that:

- we have a sled that has been expunged
- this is the first blueprint being generated since the sled was expunged
- so the parent blueprint still shows any external DNS zones on that sled as not-yet-expunged
- we want to be able to use the external IP addresses of those external DNS zones _in this blueprint generation_ to place a new external DNS zone

Is that right?

What if we relaxed that last constraint?  That'd mean that you'd have to go through two rounds of blueprint generation to replace an external DNS zone: one that expunges zones on the expunged sled plus one that adds replacement zones elsewhere.  That seems okay to me and preferable to this filter.  I don't think this is a big deal but this filter is essentially encoding some of the expunge logic here, even though these zones _aren't_ yet expunged.

Also, do we not have the same problem for Nexus zones' external IPs?

Reply:
I think I was indeed assuming that last constraint (one generation of planning) implicitly. I agree that this would be better if we allowed multiple planning passes, and will be happy to patch it to do so.

I don't know about Nexus, I will look into it while I'm preparing that patch.

Reply:
So... removing the problematic filter and relying on replanning does work fine for external DNS, but is currently broken for Nexus. If we expunge a sled, replan, and then replan again (with no changes), we get:
```
planning input contains unexpected external IP (IP not found in parent blueprint): OmicronZoneExternalIpEntry { zone_id: 205335f2-cb2a-4b73-a3ff-519f0c4db30d (service), ip: Floating(OmicronZoneExternalFloatingIp { id: 6605327a-a49e-4291-be4f-02b767b712a2 (external_ip), ip: 192.0.2.2 }) }
```
This is coming from the `Check the planning input` block around line 193. If I comment out that block the replanning happens successfully, but I'm not exactly clear what the implications of that are.

Reply:
It looks like existing tests got around this by calling `update_network_resources_from_blueprint` manually with an earlier blueprint. But that won't work in the usual case, right?

Reply:
Oh, also, sorry got distracted by the error and forgot to chime in on this bit:

> What if we relaxed that last constraint? That'd mean that you'd have to go through two rounds of blueprint generation to replace an external DNS zone: one that expunges zones on the expunged sled plus one that adds replacement zones elsewhere. That seems okay to me and preferable to this filter. I don't think this is a big deal but this filter is essentially encoding some of the expunge logic here, even though these zones aren't yet expunged.
>
> Also, do we not have the same problem for Nexus zones' external IPs?

+1 on requiring two rounds of blueprint generation. I didn't run into this with Nexus because I never tested replacing a Nexus without spare service IPs laying around; we could expunge a Nexus and add in the same blueprint, but the added one would pick a different, free service IP. I'm not sure what happens if we try to expunge a Nexus with zero spare IPs - I'll try that.

Reply:
Commit 8fe56d513 implements the 2-round planning discussed above.